File: align_parser/src/align_parsing/AlignFileParser.java
package align_parsing;

import com.google.gson.Gson;

import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.LineNumberReader;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

class AlignFileParser {
    public static HashMap<Integer, ParsedSurah> parseFile(String alignFilePath, String textFilePath) throws FileNotFoundException {
        Gson gson = new Gson();
        HashMap<Integer, ParsedSurah> surahs = new HashMap<>();
        LineNumberReader lineNumberReader = new LineNumberReader(new FileReader(textFilePath));
        List<SurahEntry> entries = Arrays.asList(
                gson.fromJson(new FileReader(alignFilePath), SurahEntry[].class));
        // Sort the file
        Comparator<SurahEntry> surahEntryComparator = (x, y) -> {
            boolean equalSurah = x.surah == y.surah;
            return equalSurah ? Integer.compare(x.ayah, y.ayah) : Integer.compare(x.surah, y.surah);
        };
        entries = entries.stream().sorted(surahEntryComparator).collect(Collectors.toList());

        for (SurahEntry entry : entries) {
            List<Segment> segments = Segment.makeSegmentsFromIntArray(entry.segments);
            ParsedAyah ayah = null;
            try {
                ayah = new ParsedAyah(entry.ayah, segments, entry.stats, lineNumberReader.readLine());
            } catch (IOException e) {
                e.printStackTrace();
                Runtime.getRuntime().exit(-1);
            }

            if (surahs.containsKey(entry.surah)) {
                ParsedSurah surah = surahs.get(entry.surah);
                surah.addAyah(ayah);
            } else {
                ParsedSurah surah = new ParsedSurah(entry.surah);
                surah.addAyah(ayah);
                surahs.put(entry.surah, surah);
            }
        }

        return surahs;
    }
}

File: align_parser/src/align_parsing/AlignFileParser.kt
package align_parsing

import com.google.gson.Gson
import java.io.FileNotFoundException
import java.io.FileReader
import java.io.IOException
import java.io.LineNumberReader
import java.util.*
import kotlin.Comparator

internal object AlignFileParser {
    @Throws(FileNotFoundException::class)
    fun parseFile(alignFilePath: String, textFilePath: String): Map<Int, Map<Int, ParsedAyah>> {
        val gson = Gson()
        val surahs = HashMap<Int, HashMap<Int, ParsedAyah>>()
        val lineNumberReader = LineNumberReader(FileReader(textFilePath))
        var entries = listOf(*gson.fromJson(FileReader(alignFilePath), Array<SurahEntry>::class.java))
        // Sort the file
        val surahEntryComparator = Comparator<SurahEntry> { x, y ->
            val equalSurah = x.surah == y.surah
            if (equalSurah) x.ayah.compareTo(y.ayah) else x.surah.compareTo(y.surah)
        }
        entries = entries.sortedWith(surahEntryComparator)

        for (entry in entries) {
            val segments = Segment.makeSegmentsFromIntArray(entry.segments)
            var ayah: ParsedAyah? = null
            try {
                ayah = ParsedAyah(entry.surah, entry.ayah, segments, entry.stats, lineNumberReader.readLine())
            } catch (e: IOException) {
                e.printStackTrace()
                Runtime.getRuntime().exit(-1)
            }

            if (!surahs.containsKey(entry.surah)) {
                surahs[entry.surah] = hashMapOf()
            }

            if (surahs[entry.surah]!!.containsKey(entry.ayah)) {
                throw IllegalStateException("Each Ayah has only ONE align entry")
            }

            surahs[entry.surah]!![entry.ayah] = ayah!!
        }

        return surahs
    }
}
